pkg/cmd/infra/deployer/deployer.go

```diff
@@ -77,88 +76,124 @@ func NewCommandDeployer(name string) *cobra.Command {
+	strategyFor func(config *deployapi.DeploymentConfig) (strategy.DeploymentStrategy, error)
+	// getDeployment finds the named deployment.
+	getDeployment func(namespace, name string) (*kapi.ReplicationController, error)
+	// getControllers finds all controllers in namespace.
+	getControllers func(namespace string) (*kapi.ReplicationControllerList, error)
+	// scaler is used to scale replication controllers.
+	scaler kubectl.Scaler
+}
+
+// Deploy starts the deployment process for deploymentName.
+func (d *Deployer) Deploy(namespace, deploymentName string) error {
+	// Look up the new deployment.
+	deployment, err := d.getDeployment(namespace, deploymentName)
 	if err != nil {
-		return err
+		return fmt.Errorf("couldn't get deployment %s/%s: %v", namespace, deploymentName, err)
 	}
 
+	// Decode the config from the deployment.
 	config, err := deployutil.DecodeDeploymentConfig(deployment, latest.Codec)
 	if err != nil {
 		return fmt.Errorf("couldn't decode DeploymentConfig from deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
 	}
 
-	var strategy strategy.DeploymentStrategy
-
-	switch config.Template.Strategy.Type {
-	case deployapi.DeploymentStrategyTypeRecreate:
-		strategy = recreate.NewRecreateDeploymentStrategy(kClient, latest.Codec)
-	case deployapi.DeploymentStrategyTypeRolling:
-		recreate := recreate.NewRecreateDeploymentStrategy(kClient, latest.Codec)
-		strategy = rolling.NewRollingDeploymentStrategy(deployment.Namespace, kClient, latest.Codec, recreate)
-	default:
-		return fmt.Errorf("unsupported strategy type: %s", config.Template.Strategy.Type)
+	// Get a strategy for the deployment.
+	strategy, err := d.strategyFor(config)
+	if err != nil {
+		return err
 	}
 
-	return strategy.Deploy(deployment, oldDeployments)
-}
-
-// getDeployerContext finds the target deployment and any deployments it considers to be prior to the
-// target deployment. Only deployments whose LatestVersion is less than the target deployment are
-// considered to be prior.
-func getDeployerContext(controllerGetter replicationControllerGetter, namespace, deploymentName string) (*kapi.ReplicationController, []*kapi.ReplicationController, error) {
-	var err error
-	var newDeployment *kapi.ReplicationController
-	var newConfig *deployapi.DeploymentConfig
-
-	// Look up the new deployment and its associated config.
-	if newDeployment, err = controllerGetter.Get(namespace, deploymentName); err != nil {
-		return nil, nil, err
+	// New deployments must have a desired replica count.
+	desiredReplicas, hasDesired := deployutil.DeploymentDesiredReplicas(deployment)
+	if !hasDesired {
+		return fmt.Errorf("deployment %s has no desired replica count", deployutil.LabelForDeployment(deployment))
 	}
 
-	if newConfig, err = deployutil.DecodeDeploymentConfig(newDeployment, latest.Codec); err != nil {
-		return nil, nil, err
+	// Find all controllers in order to pick out the deployments.
+	controllers, err := d.getControllers(namespace)
```

Comment:
It'd be better to retrieve only the controllers for the deployment config by filtering with the 'deploymentconfig' label.